OpenSim/Common/Component.cpp

```diff
@@ -838,26 +836,74 @@ void Component::constructOutputForStateVariable(const std::string& name)
+        auto& prop = updPropertyByIndex(i);
+        // check if property is of type Object
+        if (prop.isObjectProperty()) {
+            // a property is a list so cycle through its contents
+            for (int j = 0; j < prop.size(); ++j) {
+                Object& obj = prop.updValueAsObject(j);
+                // if the object is a Component mark it
+                if (Component* comp = dynamic_cast<Component*>(&obj) ) {
+                    markAsSubcomponent(comp);
+                }
+                else {
+                    // otherwise it may be a Set (of objects), and
+                    // would prefer to do something like this to test:
+                    //  Set<Component>* objects = dynamic_cast<Set<Component>*>(obj)
+                    // Instead we can see if the object has a property called
+                    // "objects" which is a PropertyObjArray used by Set<T>.
+                    // knowing the object Type is useful for debugging
+                    // and it could be used to strengthen the test (e.g. scan  
+                    // for "Set" in the type name). 
+                    std::string objType = obj.getConcreteClassName();
+                    if (obj.hasProperty("objects")) {
+                        // get the PropertyObjArray if the object has one
+                        auto& objectsProp = obj.updPropertyByName("objects");
+                        // loop over the objects in the PropertyObjArray
+                        for (int k = 0; k < objectsProp.size(); ++k) {
+                            Object& obj = objectsProp.updValueAsObject(k);
+                            // if the object is a Component mark it
+                            if (Component* comp = dynamic_cast<Component*>(&obj) )
+                                markAsSubcomponent(comp);
+                        } // loop over objects and mark it if it is a component
+                    } // end if property is a Set with "objects" inside
+                } // end of if/else property value is an Object or something else
+            } // loop over the property list
+        } // end if property is an Object
+    } // loop over properties
+}
+
 // Include another Component as a subcomponent of this one. If already a
 // subcomponent, it is not added to the list again.
-void Component::addComponent(Component* component)
+void Component::markAsSubcomponent(Component* component)
 {
-    // Only add if the Component is not already a part of the model
+    // Only add if the component is not already a part of this Component
     // So, add if empty
     if ( _components.empty() ){
         _components.push_back(component);
     }
-    else{ //otherwise check that it isn't apart of the component already        
+    else{ //otherwise check that it isn't a part of the component already
         SimTK::Array_<Component *>::iterator it =
             std::find(_components.begin(), _components.end(), component);
```

Comment:
This is implemented with an O(n^2) algorithm which could be very slow if many subcomponents were added to a particular component (because the list being searched keeps growing). I assume you aren't expecting long lists here so this is probably fine, but there should be some hard-to-miss warning comments here pointing out the potential for disaster with this algorithm.